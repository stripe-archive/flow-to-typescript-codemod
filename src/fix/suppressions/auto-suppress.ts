import { Diagnostic, Project, SourceFile, ts } from 'ts-morph'
import { FixCommandState, getDiagnostics } from '../state'
import { FixCommandCliArgs } from '../../cli/arguments'
import { logger } from '../../runner/logger'
import { isDiagnosticSuppressible } from '../insuppressible-errors'
import { diagnosticToDescription } from './diagnostic-to-description'
import { CommentToMake, CommentType } from './shared'

interface Metrics {
  suppressed: number
}

function makeComment({
  commentType,
  jiraSlug,
  diagnostics,
  annotation = '@ts-expect-error',
}: {
  commentType: CommentType
  jiraSlug: string
  diagnostics: Diagnostic[]
  annotation?: string
}): string {
  // If jira slug was not specified, don't include anything
  const fullJiraSlug = jiraSlug === '' ? '' : ` [${jiraSlug}]`

  const commentText = `${annotation}${fullJiraSlug} - ${diagnostics
    .map((diagnostic) => diagnosticToDescription(diagnostic))
    .join(' | ')}`

  if (commentType === CommentType.Jsx) {
    return `{ /* ${commentText} */}\n`
  }
  return `// ${commentText}\n`
}

/**
 * Add suppression comments for errors in the file
 */
function addSuppressionsInFile(
  metrics: Metrics,
  positions: Record<number, CommentToMake>,
  sourceFile: SourceFile,
  project: Project,
  { jiraSlug, useIgnore }: FixCommandCliArgs
) {
  let addedLength = 0
  for (const { position, commentType, diagnostics } of Object.values(
    positions
  )) {
    const isInsuppressible = diagnostics.find((error) => {
      const isInsuppressible = !isDiagnosticSuppressible(error)
      if (isInsuppressible) {
        logger.error(
          `Found an insuppressible error. Please fix manually:
            ${project.formatDiagnosticsWithColorAndContext([error])}`
        )
      }
      return isInsuppressible
    })
    if (!isInsuppressible) {
      const comment = makeComment({
        commentType,
        diagnostics,
        jiraSlug,
        annotation: useIgnore ? '@ts-ignore' : '@ts-expect-error',
      })
      const insertPos = position + addedLength
      sourceFile.insertText(insertPos, comment)
      addedLength += comment.length
      metrics.suppressed += 1
    }
  }
}

type FileWriter = (file: SourceFile) => void
const defaultWriter = (file: SourceFile) => file.saveSync()

export async function autoSuppressErrors(
  { argv, project }: FixCommandState,
  writeFile: FileWriter = defaultWriter
) {
  const diagnosticsByFile: Map<
    SourceFile,
    Record<number, CommentToMake>
  > = new Map()
  logger.info('Suppressing errors..')
  const diagnostics = getDiagnostics(project)

  diagnostics.forEach((error) => {
    const sourceFile = error.getSourceFile()

    if (!sourceFile) {
      return
    }

    const errorStart = error.getStart()
    if (!errorStart) {
      return
    }

    const node = sourceFile.getDescendantAtPos(errorStart)
    const errorStartLine = node?.getStartLinePos(true)
    if (errorStartLine == null) {
      return
    }

    // Find out which comment we should make
    const errorLine = sourceFile.getDescendantAtPos(errorStartLine)
    const commentsForFile = diagnosticsByFile.get(sourceFile) ?? {}
    const commentsForLine = commentsForFile[errorStartLine] ?? {
      position: errorStartLine,
      commentType: [
        ts.SyntaxKind.JsxElement,
        ts.SyntaxKind.JsxFragment,
        ts.SyntaxKind.JsxExpression,
        ts.SyntaxKind.JsxClosingElement,
      ].some((elementType) => errorLine?.getParentIfKind(elementType) != null)
        ? CommentType.Jsx
        : CommentType.Standard,
      diagnostics: [],
    }

    commentsForLine.diagnostics.push(error)
    commentsForFile[errorStartLine] = commentsForLine
    diagnosticsByFile.set(sourceFile, commentsForFile)
  })

  const metrics: Metrics = {
    suppressed: 0,
  }

  for (const [sourceFile, positions] of diagnosticsByFile) {
    addSuppressionsInFile(metrics, positions, sourceFile, project, argv)

    try {
      writeFile(sourceFile)
    } catch (e) {
      logger.warn(
        `Error when saving suppressed source file. Ensure that node_modules is not being type checked by your TSConfig. Error: ${e}.`
      )
    }
  }

  logger.complete(
    `Suppressed ${metrics.suppressed} errors across ${diagnosticsByFile.size} files.`
  )
}
