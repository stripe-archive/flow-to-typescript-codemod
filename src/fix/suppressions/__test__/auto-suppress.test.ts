import { autoSuppressErrors } from '../auto-suppress'
import {
  createOutputRecorder,
  getTestFixState,
} from '../../../convert/utils/testing'

jest.mock('../../../runner/migration-reporter')
jest.mock('../../../runner/logger')

describe('autoSuppressErrors', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('adds suppressions to suppressable errors', async () => {
    const [results, recordTestResult] = createOutputRecorder()

    await autoSuppressErrors(
      getTestFixState({
        tsProps: false,
        autoSuppressErrors: false,
        generateReport: false,
        jiraSlug: '',
        useIgnore: false,
        removeUnused: false,
        config: './src/fix/suppressions/__test__/tsconfig.json',
        format: 'stdout',
        silent: false,
        output: '',
        autoImport: false,
        fixTypeExports: false,
      }),
      recordTestResult
    )

    expect(results['test-input.ts']).toMatchSnapshot()
  })

  it('does not add if run twice', async () => {
    const [results, recordTestResult] = createOutputRecorder()
    const state = getTestFixState({
      tsProps: false,
      autoSuppressErrors: false,
      generateReport: false,
      jiraSlug: '',
      useIgnore: false,
      removeUnused: false,
      config: './src/fix/suppressions/__test__/tsconfig.json',
      format: 'stdout',
      silent: false,
      output: '',
      autoImport: false,
      fixTypeExports: false,
    })

    await autoSuppressErrors(state, recordTestResult)

    await autoSuppressErrors(state, recordTestResult)

    expect(results['test-input.ts']).toMatchSnapshot()
  })
})
