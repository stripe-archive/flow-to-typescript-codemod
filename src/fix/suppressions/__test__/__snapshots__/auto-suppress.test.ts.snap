// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autoSuppressErrors adds suppressions to suppressable errors 1`] = `
"// Testing removing unused

// @ts-expect-error - TS2322 - Type 'number' is not assignable to type 'string'. | TS2451 - Cannot redeclare block-scoped variable 'error1'.
const error1: string = 0

// Line above suppression
// @ts-expect-error not actually error
export const foo = () => {
  console.log(
    // Other comment
    'foo',
    // @ts-expect-error - TS2345 - Argument of type 'unknown' is not assignable to parameter of type '{ customer: string; ephemeralKey: string; } | null'.
    'bar',
    'baz'
  )

// @ts-expect-error - TS2322 - Type 'number' is not assignable to type 'string'.
  const error2: string = 0

  /* @ts-expect-error not actually error */
  return 'bar'
}

// @ts-expect-error - TS2322 - Type 'number' is not assignable to type 'string'. | TS2451 - Cannot redeclare block-scoped variable 'error1'.
const error1: string = 0
"
`;

exports[`autoSuppressErrors does not add if run twice 1`] = `
"// Testing removing unused

// @ts-expect-error - TS2322 - Type 'number' is not assignable to type 'string'. | TS2451 - Cannot redeclare block-scoped variable 'error1'.
const error1: string = 0

// Line above suppression
// @ts-expect-error not actually error
export const foo = () => {
  console.log(
    // Other comment
    'foo',
    // @ts-expect-error - TS2345 - Argument of type 'unknown' is not assignable to parameter of type '{ customer: string; ephemeralKey: string; } | null'.
    'bar',
    'baz'
  )

// @ts-expect-error - TS2322 - Type 'number' is not assignable to type 'string'.
  const error2: string = 0

  /* @ts-expect-error not actually error */
  return 'bar'
}

// @ts-expect-error - TS2322 - Type 'number' is not assignable to type 'string'. | TS2451 - Cannot redeclare block-scoped variable 'error1'.
const error1: string = 0
"
`;
